### PR TITLE
*: Support show status statement

### PR DESCRIPTION
**Pull request by zimulala**
#525

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -285,6 +283,35 @@ func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, err
 	return value, nil
 }
 
+// GetGlobalStatusVar implements RestrictedSQLExecutor.GetGlobalStatusVar interface.
```
Comment (shenli) [inline session.go]: implements GlobalVarAccessor.GetGlobalStatusVar interface.
So do the othe three functions.

File: session.go [hunk 2]
```diff
@@ -285,6 +283,35 @@ func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, err
...
+func (s *session) GetGlobalStatusVar(ctx context.Context, name string) (string, error) {
+	// TODO: get global status variables from store.
```
Comment (shenli) [inline session.go]: You can use a table in system db to store status and init this table when doing bootstrap.

File: session.go [hunk 3]
```diff
@@ -285,6 +283,35 @@ func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, err
...
+
+	return v.Value, nil
+}
+
+// SetGlobalStatusVar implements RestrictedSQLExecutor.SetGlobalStatusVar interface.
+func (s *session) SetGlobalStatusVar(ctx context.Context, name string, value string) error {
+	// TODO: set global status variables from store.
+	v := variable.GetStatusVar(name)
+	if v == nil {
+		return fmt.Errorf("Unknown status var: %s", name)
```
Comment (shenli) [inline session.go]: errors.Errorf?

File: session.go [hunk 4]
```diff
@@ -263,9 +263,7 @@ func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recor
 	return rs, errors.Trace(err)
 }
 
-// GetGlobalSysVar implements RestrictedSQLExecutor.GetGlobalSysVar interface.
-func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, error) {
-	sql := fmt.Sprintf(`SELECT VARIABLE_VALUE FROM %s.%s WHERE VARIABLE_NAME="%s";`, mysql.SystemDB, mysql.GlobalVariablesTable, name)
+func (s *session) getGlobalVar(ctx context.Context, name, sql string) (string, error) {
```
Comment (shenli) [inline session.go]: This is a little confused. The interface is not clear enough. And the name parameter is not used.

**Comment by siddontang** [comment]
how do we add our own status variables and show them?

File: session.go (hunk 4 above)
Comment (zimulala) [inline session.go]: I think it can be used by GetGlobalStatusVar later ?

File: session.go (hunk 3 above)
Comment (zimulala) [inline session.go]: got it.

File: session.go (hunk 2 above)
Comment (zimulala) [inline session.go]: yep, I will do it in the next pr.

File: session.go (hunk 1 above)
Comment (zimulala) [inline session.go]: you means `SetGlobalStatusVar` ,`GetGlobalSysVar` and `SetGlobalSysVar` ?

**Comment by zimulala** [comment]
add them to the corresponding global status variables or session status variables

File: plan/plans/show.go [hunk 5]
```diff
@@ -402,6 +406,50 @@ func (s *ShowPlan) fetchShowVariables(ctx context.Context) error {
...
+		if !match {
+			continue
+		}
+
+		var value string
+		if !s.GlobalScope {
+			sv, ok := sessionVars.StatusVars[v.Name]
+			if ok {
+				value = sv
+			}
```
Comment (shenli) [inline plan/plans/show.go]: https://dev.mysql.com/doc/refman/5.7/en/show-status.html
With a SESSION modifier, the statement displays the status variable values for the current connection. If a variable has no session value, the global value is displayed.

File: session.go [hunk 6]
```diff
@@ -263,9 +263,7 @@ func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recor
 	return rs, errors.Trace(err)
 }
 
-// GetGlobalSysVar implements RestrictedSQLExecutor.GetGlobalSysVar interface.
-func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, error) {
-	sql := fmt.Sprintf(`SELECT VARIABLE_VALUE FROM %s.%s WHERE VARIABLE_NAME="%s";`, mysql.SystemDB, mysql.GlobalVariablesTable, name)
+func (s *session) getExecRet(ctx context.Context, sql string) (string, error) {
```
Comment (shenli) [inline session.go]: Obviously, there are a few constraints for the sql (The result has exactly one row and one column, the value is a string).
Pick a better self-explain name or add some comments.

File: session.go [hunk 7]
```diff
@@ -276,7 +274,7 @@ func (s *session) GetGlobalSysVar(ctx context.Context, name string) (string, err
 		return "", errors.Trace(err)
 	}
 	if row == nil {
-		return "", fmt.Errorf("Unknown sys var: %s", name)
+		return "", errors.New("Unknown sys var")
```
Comment (shenli) [inline session.go]: You can output this log outside this function. And add var name in the log.

File: plan/plans/show.go (hunk 5 above)
Comment (zimulala) [inline plan/plans/show.go]: but it in mysql operation result is as follows:
```
mysql> show session Status like 'Comp%';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| Compression   | OFF   |
+---------------+-------+
1 row in set (0.00 sec)

mysql> show global Status like 'Comp%';
Empty set (0.00 sec)
```

File: session.go (hunk 7 above)
Comment (siddontang) [inline session.go]: errors.Errorf may be better .

File: plan/plans/show.go (hunk 5 above)
Comment (shenli) [inline plan/plans/show.go]: mysql> show session Status like 'Comp%';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| Compression   | OFF   |
+---------------+-------+
1 row in set (0.00 sec)

mysql> show global Status like 'Comp%';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| Compression   | OFF   |
+---------------+-------+
1 row in set (0.00 sec)

mysql>

File: session.go (hunk 6 above)
Comment (zimulala) [inline session.go]: Do you have some suggestions?

File: plan/plans/show.go (hunk 5 above)
Comment (zimulala) [inline plan/plans/show.go]: my mysql version is 5.7.9
`If a variable has no global value, the session value is displayed.`
The sentence mislead me.
Your advice is better, I will modify it right now.

**Comment by zimulala** [comment]
PTAL @siddontang  @shenli

File: util/mock/context.go [hunk 8]
```diff
@@ -57,21 +56,38 @@ func (c *Context) FinishTxn(rollback bool) error {
...
+		return "", nil
+	}
+	return v.Value, nil
+}
+
+// SetGlobalStatusVar implements GlobalVarAccessor SetGlobalStatusVar interface.
+func (c *Context) SetGlobalStatusVar(ctx context.Context, name string, value string) error {
+	v := variable.GetStatusVar(name)
+	if v == nil {
+		return fmt.Errorf("Unknown status var: %s", name)
```
Comment (siddontang) [inline util/mock/context.go]: errors.Errorf

File: util/mock/context.go [hunk 9]
```diff
@@ -57,21 +56,38 @@ func (c *Context) FinishTxn(rollback bool) error {
...
 
-// SetGlobalSysVar implements GlobalSysVarAccessor SetGlobalSysVar interface.
+// SetGlobalSysVar implements GlobalVarAccessor SetGlobalSysVar interface.
 func (c *Context) SetGlobalSysVar(ctx context.Context, name string, value string) error {
-	sysvars := variable.SysVars
-	v, ok := sysvars[strings.ToLower(name)]
-	if !ok {
+	v := variable.GetSysVar(name)
+	if v == nil {
 		return fmt.Errorf("Unknown sys var: %s", name)
```
Comment (siddontang) [inline util/mock/context.go]: errors.Errorf

File: plan/plans/show.go [hunk 10]
```diff
@@ -402,6 +407,86 @@ func (s *ShowPlan) fetchShowVariables(ctx context.Context) error {
...
+
+func getGlobalStatusVar(ctx context.Context, sessionVars *variable.SessionVars,
+	globalVars variable.GlobalVarAccessor, name string) (string, error) {
+
+	value, err := globalVars.GetGlobalStatusVar(ctx, name)
+	if err == nil {
+		return value, nil
+	}
+
+	if terror.UnknownStatusVar.Equal(err) {
```
Comment (siddontang) [inline plan/plans/show.go]: check error first

File: util/mock/context.go (hunk 8 above)
Comment (zimulala) [inline util/mock/context.go]: got it

File: plan/plans/show.go (hunk 10 above)
Comment (zimulala) [inline plan/plans/show.go]: it checks in line 429

File: util/mock/context.go [hunk 11]
```diff
@@ -57,22 +57,39 @@ func (c *Context) FinishTxn(rollback bool) error {
...
-// SetGlobalSysVar implements GlobalSysVarAccessor SetGlobalSysVar interface.
+// SetGlobalSysVar implements GlobalVarAccessor SetGlobalSysVar interface.
 func (c *Context) SetGlobalSysVar(ctx context.Context, name string, value string) error {
-	sysvars := variable.SysVars
-	v, ok := sysvars[strings.ToLower(name)]
-	if !ok {
-		return fmt.Errorf("Unknown sys var: %s", name)
+	v := variable.GetSysVar(name)
+	if v == nil {
+		return terror.UnknownStatusVar.Gen("unknown sys variable: %s", name)
```
Comment (shenli) [inline util/mock/context.go]: Need a new error like terror.UnknownSystemVar?

Comment (zimulala) [inline util/mock/context.go]: yep.

File: terror/terror.go [hunk 12]
```diff
@@ -52,6 +57,12 @@ const (
 	CodeNoDataForHandle
 )
 
+// Variable error codes.
+const (
+	CodeUnknownStatusVar ErrCode = iota + 1
```
Comment (shenli) [inline terror/terror.go]: CodeCommitNotInTransaction ErrCode = iota + 1
CodeUnknownStatusVar ErrCode = iota + 1

ErrorCode should be unique? @coocood

Comment (coocood) [inline terror/terror.go]: ErrorCode should be unique in the same class, but doesn't have to be globally unique.

**Comment by zimulala** [comment]
PTAL @siddontang  @shenli

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
PTAL @siddontang

File: plan/plans/show.go (hunk 10 above)
Comment (siddontang) [inline plan/plans/show.go]: no need to return err if it is not UnknownStatusVar?

**Comment by siddontang** [comment]
LGTM

**Pull request merged**